make/common/NativeCompilation.gmk

```diff
@@ -1143,6 +1149,19 @@ define SetupNativeCompilationBody
       endif
     endif
 
+    # Debuginfo of ASM objects always embeds the absolute object path,
+    # as ASM debuginfo paths do not get prefix mapped.
+    # So for reproducible builds use relative paths to ensure a reproducible
+    # debuginfo and libs, when creating debug symbols.
+    ifeq ($$(ENABLE_REPRODUCIBLE_BUILD), true)
+      ifeq ($(call isTargetOs, linux), true)
+        ifeq ($$($1_COMPILE_WITH_DEBUG_SYMBOLS), true)
+          $1_LINK_OBJS_RELATIVE := true
```

Comment:
I realize this PR has already been integrated, but I have some questions about this. The comment talk about "ASM" objects (I assume that means output of assembly files), but the code change is not restricted to assembly files. 

Afaict, this change means that all builds on linux with debug symbols and reproducible builds will use relative paths! This is perhaps the most common compilation scenario, and it would mean that our efforts to keep a CWD neutral command line is basically in vain. :-(

Or can anyone (@erikj79, @andrew-m-leonard) explain to me why this would not be the case?

Reply:
You are correct for the linking command line. All the compilation command lines are still handled with flags instead.

Reply:
What is CWD ?

Reply:
Current working dir !

Reply:
@magicus 
So I did think about this, and the alternative I think would be to hard code a list of which object files are the output from assembly files, of which there is about a dozen. Unless we could capture somehow the list of assembled files as we go ...?
As @erikj79 points out this is just for the "linking". As the problem is the assembly file objects don't have the equivalent of debug prefix mapping, and so when stripped the full file path of the object file ends up in the debuginfo file, and results in the libjvm.so non-determinism due to different debug CRC.

Reply:
It would certainly be possible to identify the objects created from assembly source files based on file extensions, but it wouldn't help. The command line would still be dependent on the current working directory (correctly guessed!) if just one of the argument files is relative.

The property Magnus is trying to preserve is to be able to copy a command line from the build log and run it in whichever directory you happen to be in, without having to figure out what the CWD was for that particular command. We worked hard to try to figure out ways of preserving that as much as possible, but as you say, if we want to reproducibility, then we have to give it up for link command lines where assembly source files are included.

Now I realize that you aren't even checking if the link unit contains any objects from assembly files. I think we should be able to deduce that by filtering on the extension in the source file list. That would miss assembly input in EXTRA_OBJS, so perhaps we need an explicit parameter to SetupNativeCompilation for cases where this would happen. I think it applies to gtest, where we add all the object files from libjvm.

Reply:
One way would be to check if we have either assembly source files, or anything in EXTRA_OBJ, and if so, do the relative linking. That would at least bring down collateral damage significantly, since the majority of libs have neither.

It will still hurt the most important libs (like hotspot) but unless we can't find a way to get this working in some other way, it seems we'll have to accept that.

@andrew-m-leonard Do you know if there is any bug on gcc tracking their inability to set debug prefix mapping for assembly files?

Reply:
Actually, I think that the GNU assembler supports debug prefix mapping. See e.g. [this bug report](https://github.com/ocaml/ocaml/issues/10770), stating: "With GNU tools, to enable debug prefix map on C source we pass -fdebug-prefix-map to cc and --debug-prefix-map to as."

@andrew-m-leonard Can you check if we can pass `--debug-prefix-map` to the assembler if on the gcc toolchain, and thus get rid of the relative linking? (And oh, speaking of that, the if statement here checks for target OS. It should, most likely, have checked for toolchain.)

Reply:
@magicus thank you for looking at this, in hindsight I should have pinged you to review this as well.

So a few thoughts, first one thing that crossed my mind which is the ifeq logic here should instead test ALLOW_ABSOLUTE_PATHS_IN_OUTPUT, exactly like the setting of -fdebug-prefix-map and -ffile-macro-prefix already do. In fact just changing to using that maybe sufficient? ie.only use relative paths here when ALLOW_ABSOLUTE_PATHS_IN_OUTPUT is false, ie. it is a "release" type build and not a developer/debug build, where CWD cmdlines are preferable. What's your thought on that?

I did search google & bugzilla for any gcc bug, but didn't find anything. However, i've just found this issue: https://github.com/gcc-mirror/gcc/commit/d12153046816f955e74943af7089d30de6a00e19
which maybe related, although looks to do the opposite of what we want, although that maybe just because I don't understand it! However, it is the .file <path> dwarf value for the assembly stripped debuginfo that has the full object path that isn't getting stripped, and I think we do use -flto. This fix is in gcc 11.2, i'm testing using gcc 10.3.
I'm going to do two things, recreate the problem again, double check the assembly file compile options, with gcc 10.3, then try again with gcc 11.2..

Reply:
@magicus just saw your above post, interesting, i'll add I think gcc -v to my above testing, to see what gcc is passing through to the actual compile, apparently using verbose removes anything that gets removed for assembly, ie.anything that is not relavent

Reply:
I did test with --debug-prefix-map on the assembly cmdline, it just seemed to be ignored, but I will try again